### PR TITLE
Most of the fields in "partial rerun merge" events should be optional

### PR DESCRIPTION
**Pull request by corva-alex-s**
### Rationale
During testing on QA we discovered that some of the fields might be missing or null, in such cases current SDK logic would not recognize them as "partial rerun merge" events and will fail to process them properly.

### Changes
changed most of the fields to be optional. The only fields required are those that we need to setup cache.

[JIRA ticket](https://corvaqa.atlassian.net/browse/DEVC-570)

#### TODO
- [x] Update CHANGELOG.md

### CODE REVIEW

File: src/corva/handlers.py [hunk 1]
```diff
@@ -576,6 +576,11 @@ def _merge_events(
     Only "scheduled" and "stream" type of apps can be processed here.
     If somehow any other type is passed - raise an exception
     """
+    if not isinstance(
```
Comment (vladbagmet) [inline src/corva/handlers.py:579]: Why should we run functionality to merge event if event does not have any data to merge?

File: src/corva/models/merge/merge.py [hunk 2]
```diff
@@ -25,23 +27,23 @@ class PartialRerunMergeEvent(CorvaBaseEvent):
         run_until: run until
     """
 
-    event_type: EventType
-    partial_well_rerun_id: int
-    rerun_mode: RerunMode
-    start: int
-    end: int
+    event_type: Optional[EventType]
```
Comment (vladbagmet) [inline src/corva/models/merge/merge.py:30]: This is a mandatory field, not optional. They have to send it. If they do not send it this is a problem that should be communicated.

File: src/corva/models/merge/merge.py [hunk 3]
```diff
@@ -25,23 +27,23 @@ class PartialRerunMergeEvent(CorvaBaseEvent):
...
+    start: Optional[int]
+    end: Optional[int]
     asset_id: int
     rerun_asset_id: int
     app_stream_id: int
     rerun_app_stream_id: int
-    version: int
-    app_id: int
-    app_key: str
+    version: Optional[int]
```
Comment (vladbagmet) [inline src/corva/models/merge/merge.py:39]: I think this field can not be optional as well. It's part of the mechanism of backwards-compatibility when the shape of data is changed.

File: tests/unit/test_partial_rerun_merge_app.py [hunk 4]
```diff
@@ -121,3 +121,44 @@ def partial_rerun_merge_app(event, api, asset_cache, rerun_asset_cache):
...
+
+    @corva.scheduled(merge_events=True)
+    def scheduled_app(event, api, cache):
+        pytest.fail("Scheduled app was unexpectedly called!")
+
+    @corva.partial_rerun_merge
+    def partial_rerun_merge_app(event, api, asset_cache, rerun_asset_cache):
+        return True
+
+    event_with_missing_parameter: dict = deepcopy(RAW_EVENT)
```
Comment (vladbagmet) [inline tests/unit/test_partial_rerun_merge_app.py:140]: You can avoid using additional module by doing something like this
` event_with_missing_parameter = dict(RAW_EVENT)`

File: tests/unit/test_partial_rerun_merge_app.py [hunk 5]
```diff
@@ -121,3 +121,44 @@ def partial_rerun_merge_app(event, api, asset_cache, rerun_asset_cache):
...
+    event_with_missing_parameter["data"]["end"] = None
+    partial_merge_app_response = scheduled_app(event_with_missing_parameter, context)[0]
+    assert partial_merge_app_response is True
+
+
+def test_merge_events_parameter_ignored_for_partial_rerun_merge(context):
+    """
+    merge_events parameter is ignored when incoming aws_event is not a list
```
Comment (vladbagmet) [inline tests/unit/test_partial_rerun_merge_app.py:148]: Good case to test. But where in your test do you actually test it?

**Review by vladbagmet — COMMENTED** [review]
@corva-alex-s, I left some comments. The main question is why do we run functionality to merge event for event which does not have any data to merge?

File: src/corva/models/merge/merge.py (hunk 2 above)
Comment (oleksii-symon-corva-ai) [inline src/corva/models/merge/merge.py:30]: Ideally this field should be used to discriminate between different event types, so it should not be optional
https://docs.pydantic.dev/1.10/usage/types/#nested-discriminated-unions

**Comment by oleksii-symon-corva-ai** [comment]
@vladbagmet are `HANDLERS` supposed to be empty?
https://github.com/corva-ai/python-sdk/blob/DEVC-570-be-sdk-fix-but-with-merge-events-parameter-for-slide-sheet-scheduler/src/corva/handlers.py#L42

File: tests/unit/test_partial_rerun_merge_app.py [hunk 6]
```diff
@@ -121,3 +121,45 @@ def partial_rerun_merge_app(event, api, asset_cache, rerun_asset_cache):
...
+
+    @corva.partial_rerun_merge
+    def partial_rerun_merge_app(event, api, asset_cache, rerun_asset_cache):
+        return True
+
+    partial_merge_app_response = scheduled_app(RAW_EVENT, context)[0]
+    assert partial_merge_app_response is True
+
+
+def test_partial_rerun_decorator_can_be_called_directly(context):
```
Comment (oleksii-symon-corva-ai) [inline tests/unit/test_partial_rerun_merge_app.py:164]: What is the purpose of this test?

File: src/corva/models/merge/merge.py (hunk 2 above)
Comment (oleksii-symon-corva-ai) [reply src/corva/models/merge/merge.py:30]: I see, it is not optional below

**Comment by vladbagmet** [comment]
> @vladbagmet are `HANDLERS` supposed to be empty? https://github.com/corva-ai/python-sdk/blob/DEVC-570-be-sdk-fix-but-with-merge-events-parameter-for-slide-sheet-scheduler/src/corva/handlers.py#L42

@oleksii-symon-corva-ai, handlers are initialised empty and then they are populated on the fly if we have any partial-merge decorator declared within an app 
https://github.com/corva-ai/python-sdk/blob/DEVC-570-be-sdk-fix-but-with-merge-events-parameter-for-slide-sheet-scheduler/src/corva/handlers.py#L551

File: tests/unit/test_partial_rerun_merge_app.py (hunk 4 above)
Comment (vladbagmet) [reply tests/unit/test_partial_rerun_merge_app.py:140]: You are right, for nested dicts we should use deepcopy.

File: src/corva/handlers.py (hunk 1 above)
Comment (vladbagmet) [reply src/corva/handlers.py:579]: @corva-alex-s, partial merge rerun events handler was implemented before events merging functionality. What stopped you from learning how partial merge rerun functionality works before working on events merging functionality?

**Pull request merged**